### PR TITLE
Add QueryString command support

### PR DESCRIPTION
**Pull request by ygalbel**
I created queryString command support.

The code is very similar to setHeader command.
I support multiple time the same key (`like ?key=value1&key=value2`).

The value is always url encoded.
All the code is tested.

I didn't implement suggestions.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/HttpRepl?pullRequest=448) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/HttpRepl?pullRequest=448) before we can accept your contribution.<br/><br/>:x: ygalbel [sign now](https://cla.dotnetfoundation.org/dotnet/HttpRepl?pullRequest=448)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/HttpRepl?pullRequest=448) it.</sub>

**Comment by tlmii** [comment]
@ygalbel Thanks for taking the time to submit the PR!

I'll want input from @bradygaster, @sayedihashimi, etc when we get past the long holiday weekend. 

But in the interim, wanted to just clarify the purpose of this PR.

There's two primary scenarios for querystring parameters with APIs that I'm aware of:

1. Querystring parameters used across all (or most) calls to the API. Examples would be parameters needed for specifying an API Version (e.g. the `api-version` parameter on calls to the [Azure REST API](https://docs.microsoft.com/rest/api/azure/#request-uri)) or some type of authentication token (e.g. something from [Amazon's S3 API](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)). 
2. Querystring parameters that are specific to a particular API call, acting similarly to those that are part of the path of the url (the ones we can `cd` through in the tool).
3. Querystring parameters that are in common to all (or many) calls to the API, but that you wouldn't necessarily pass the same value on each call (e.g. pagination parameters)

From my initial glance at this PR, it seems like its goal is to provide an implementation for the first scenario, since the values are stored as part of the `HttpState` object:

https://github.com/dotnet/HttpRepl/blob/32f45bf63dfd576f1d33ee6b4897ce24d2053291/src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs#L50-L59

and passed to all commands:

https://github.com/dotnet/HttpRepl/blob/32f45bf63dfd576f1d33ee6b4897ce24d2053291/src/Microsoft.HttpRepl/Commands/BaseHttpCommand.cs#L134

Is that accurate?

I'd be interested to hear what the folks from #258 (@bravecobra, @salihkarabilo, @Leon99) say about whether this would address their needs, or if they need scenarios 2 & 3 as well.

**Comment by ygalbel** [comment]
Hi,
Thanks for your answer.

As you said, my code fix case number 1.

I am working against an API that get username and password as part of queryString,
so that is the reason I need to add this kind of support, and use the same queryString in all requests.

Maybe I can add the support for other scenarios if it will be used.
That don't seems to be a lot of work.

**Comment by tlmii** [comment]
@ygalbel I'd say hold off on trying to implement the others. I think scenario 1 that you've already implemented is the most straightforward from a UX perspective and we probably want to give the others a bit more consideration. I'll take a pass at reviewing the PR shortly and let the others chime in when they get back.

@darrelmiller - correct me if I'm wrong, but there's nothing in the OpenAPI spec that would differentiate in the OpenAPI description whether a parameter was "global" (like scenario 1 or 3) vs specific to the particular path/method, right? It probably _should_ be split out as a `$ref` in the global scenarios, but that's not required _and_ it could be split out as a `$ref` in scenario 2 anyway. Is that correct?

File: src/Microsoft.HttpRepl/Commands/HelpCommand.cs
Comment (tlmii) [inline src/Microsoft.HttpRepl/Commands/HelpCommand.cs:189]: From the output of this, it looks like this is now our longest command name, so we'll need to bump `navCommandColumn` (probably bump it by 2 since the command will be longer when it is changed to `query-string`):

```console
Setup Commands:
Use these commands to configure the tool for your API server

connect        Configures the directory structure and base address of the api server
set header     Sets or clears a header for all requests. e.g. `set header content-type application/json`
set queryStringSets or clears a query string key and value. When [value] is empty the header is cleared. The value will be UrlEncoded.
```

Comment (tlmii) [inline src/Microsoft.HttpRepl/Commands/HelpCommand.cs:189]: `query-string`

File: test/Microsoft.HttpRepl.Tests/HttpStateTests.cs
Comment (tlmii) [inline test/Microsoft.HttpRepl.Tests/HttpStateTests.cs:191]: Can you lift the string literal here to a local variable at the top (probably something like `expectedPath`) so they're near the construction of the parameters for easy comparison? Same for the other test methods below.

File: test/Microsoft.HttpRepl.Tests/HttpStateTests.cs
Comment (tlmii) [inline test/Microsoft.HttpRepl.Tests/HttpStateTests.cs:183]: Same nit about `var` usage in this file as mentioned elsewhere. Note that you can use target-typed new expression here to keep the generic `Dictionary`/`IEnumerable` syntax from being so ugly:
```C#
Dictionary<string, IEnumerable<string>> queryString = new()
{
   ...
}
```

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (tlmii) [inline src/Microsoft.HttpRepl/HttpState.cs:95]: I need to give this more thought. I _think_ its more correct to include the query string by default (and force the use of the static if you want only some of the parts - as is done in `GetEffectivePathForPrompt`). But I need to dig through some of the usages first.

File: test/Microsoft.HttpRepl.Tests/HttpStateTests.cs
Comment (ygalbel) [reply test/Microsoft.HttpRepl.Tests/HttpStateTests.cs:183]: That didn't works for me.
I don't know why.
For now I will keep 
````
Dictionary<string, IEnumerable<string>> queryString = new Dictionary<string, IEnumerable<string>>()
`````

File: test/Microsoft.HttpRepl.Tests/HttpStateTests.cs
Comment (ygalbel) [reply test/Microsoft.HttpRepl.Tests/HttpStateTests.cs:191]: Done

File: src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs
Comment (ygalbel) [reply src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs:26]: Done

File: src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs:39]: Interesting. What version of VS are you using?

Comment (ygalbel) [reply src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs:39]: Version 16.8.2.
I think is the latest.

Comment (tlmii) [reply src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs:39]: Hm. Yeah I can repro that too. 16.8.2 doesn't like it, but the 16.9 previews I have on my box do. Not sure why they'd be different here. Oh well. You can leave it how it works in 16.8.2 since that's the non-preview release.

**Comment by bravecobra** [comment]
> @ygalbel Thanks for taking the time to submit the PR!
> 
> I'll want input from @bradygaster, @sayedihashimi, etc when we get past the long holiday weekend.
> 
> But in the interim, wanted to just clarify the purpose of this PR.
> 
> There's three primary scenarios for query string parameters with APIs that I'm aware of:
> 
> 1. Query string parameters used across all (or most) calls to the API. Examples would be parameters needed for specifying an API Version (e.g. the `api-version` parameter on calls to the [Azure REST API](https://docs.microsoft.com/rest/api/azure/#request-uri)) or some type of authentication token (e.g. something from [Amazon's S3 API](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)).
> 2. Query string parameters that are specific to a particular API call, acting similarly to those that are part of the path of the url (the ones we can `cd` through in the tool).
> 3. Query string parameters that are in common to all (or many) calls to the API, but that you wouldn't necessarily pass the same value on each call (e.g. pagination parameters)
> 
> From my initial glance at this PR, it seems like its goal is to provide an implementation for the first scenario, since the values are stored as part of the `HttpState` object:
> 
> https://github.com/dotnet/HttpRepl/blob/32f45bf63dfd576f1d33ee6b4897ce24d2053291/src/Microsoft.HttpRepl/Commands/SetQueryStringCommand.cs#L50-L59
> 
> and passed to all commands:
> 
> https://github.com/dotnet/HttpRepl/blob/32f45bf63dfd576f1d33ee6b4897ce24d2053291/src/Microsoft.HttpRepl/Commands/BaseHttpCommand.cs#L134
> 
> Is that accurate?
> 
> I'd be interested to hear what the folks from #258 (@bravecobra, @salihkarabilo, @Leon99) say about whether this would address their needs, or if they need scenarios 2 & 3 as well.

Yeah, I would be requiring all three cases, however that would by scripting out the api calls, not by manually typing out each call at the command-line fortunately.

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (bradygaster) [reply src/Microsoft.HttpRepl/HttpState.cs:112]: thanks!

**Comment by bradygaster** [comment]
Like @tlmii said, thanks for the PR, @ygalbel and for working through some suggestions. I'm eager to get support for QS variables so we unblock a lot of key scenarios, like _anyone using `[FromQuery]` out there_, so I'd argue getting support for items 1 and 2 would a must. 3 i'm less concerned about.

**Comment by ygalbel** [comment]
Anything I can help, or you are taking the remaining development from here?
Do you want that all the feature will be finished before you will merge this Pull Request?

**Comment by bravecobra** [comment]
Does this work with the same query string key being passed multiple times with different values to provide a list? We do have that scenario too.

**Comment by ygalbel** [comment]
@bravecobra sure,
this case (`?key=value1&key=value2`) is totally supported and tested.

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (atifaziz) [inline src/Microsoft.HttpRepl/HttpState.cs:129]:
```suggestion
                        AppendQueryToBuilder(builder, $"{WebUtility.UrlEncode(tuple.Key)}={WebUtility.UrlEncode(singleValue)}");
```

According to [“application/x-www-form-urlencoded “](https://www.w3.org/TR/html401/interact/forms.html#didx-applicationx-www-form-urlencoded):

> 1. Control names and values are escaped.

It's also best to err on the safe side. I'd also recommend adding a test for this.

Also, any considerations for the following?

> Additionally, the `allowReserved` keyword specifies whether the reserved characters `:/?#[]@!$&'()*+,;=` in parameter values are allowed to be sent as they are, or should be percent-encoded. By default, `allowReserved` is `false`, and reserved characters are percent-encoded. For example, `/` is encoded as `%2F` (or `%2f`), so that the parameter value `quotes/h2g2.txt` will be sent as `quotes%2Fh2g2.txt`

File: src/Microsoft.HttpRepl/Resources/Strings.resx
Comment (atifaziz) [inline src/Microsoft.HttpRepl/Resources/Strings.resx:440]: nit: unrelated change

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (atifaziz) [inline src/Microsoft.HttpRepl/HttpState.cs:124]: nit: extra line:

```suggestion
```

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (atifaziz) [inline src/Microsoft.HttpRepl/HttpState.cs:93]: nit: extra blank line:

```suggestion
```

**Review by atifaziz — CHANGES_REQUESTED** [review]

File: src/Microsoft.HttpRepl/Commands/HelpCommand.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/Commands/HelpCommand.cs:189]: I definitely want to pull in @bradygaster on this one. I agree with the logic of what you said - including the query-param part... but I do wonder a bit about user expectations. Does it matter if (and this is just a guess on my part) our users would be more familiar with the term "query string" being used colloquially to cover a single query parameter? It seems like `query`/`query-string` would be aliases for each other (if we add those commands) and would be things the user might try on their own, but `query-param` might not be as obvious. I'd be happy to be wrong there, though, as I certainly prefer the precise language suggested here.

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/HttpState.cs:38]: > Now if the scope of HTTP REPL is just web APIs per Open API

That's certainly a/the focus, but I wouldn't want to constrain us to that when there's a relatively easy way to support more scenarios. 

That said, I see a bit of a problem here in that the list being created here (global query parameters, sent to all subsequent HTTP commands) will be combined with per-command query parameters specified at the time of the specific HTTP command.

It would seem that the best we could do would be to guarantee 
- Order within global query parameters was maintained
- Order within command-specific query parameters was maintained
- Global query parameters are always listed... before? after? command-specific query parameters

I suspect that the situation where order is important AND the user would prefer to have a mixture of global and command-specific parameters would be exceedingly rare, but I wanted to call it out here in case anyone has any thoughts.

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/HttpState.cs:129]: re: `allowReserved`:  I'd be curious to know how often allowReserved is true on query parameters (a quick [search of APIs.guru's OpenAPI Directory](https://github.com/APIs-guru/openapi-directory/search?q=allowReserved) turned up only 1 result, though I doubt that's fully representative. My instinct says we should table that and add an issue to track supporting it later, in large part because that's a per-endpoint decision and the parameter definitions aren't available at the time this code is being called, thus making it a larger change to support it. Seems like a good candidate for its own issue/PR.

File: src/Microsoft.HttpRepl/Commands/HelpCommand.cs
Comment (atifaziz) [reply src/Microsoft.HttpRepl/Commands/HelpCommand.cs:189]: Folks, during a recent GitHub service hiccup/incident, my original comment got deleted. The support staff at GitHub were kind enough to to recover the comment body, which I posting back here for context and belong right before @tlmii reply above:

> Why not make it simpler and call it `query`? Does the string part add or help anything except more typing? It would align nicely with [Section 3, “Syntax Components”](https://tools.ietf.org/html/rfc3986#section-3) (as well as [sub-section 3.4](https://tools.ietf.org/html/rfc3986#section-3.4)) of [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3) where it's called just _query_:
> 
>     foo://example.com:8042/over/there?name=ferret#nose
>     \_/ \______________/\_________/ \_________/ \__/
>     | | | | |
>     scheme authority path query fragment
>     | _____________________|__
>     / \ / \
>     urn:example:animal:ferret:nose
> 
> If anything, `query-string` should be an alias like `dir` is for `ls`.
> 
> That said, there is another problem. The query component is opaque. The fact that it contains name-value pairs came from [URL-encoding of HTML forms](https://www.w3.org/TR/html401/interact/forms.html#h-17.13.3.4) and got adopted more widely outside the context of HTML form submission. The name-value pairs have become more conventionally known as _parameters_, e.g. [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/get#Examples) and [query parameters in Open API](https://swagger.io/docs/specification/describing-parameters/#query-parameters), so I would argue that this should technically be called `query-param`. This way, `query`/`query-string` could be reserved to parsing and setting parameters given an entire “query string”. In other words, doing `set query foo=1&bar=2` would be equivalent to doing `set query-param foo 1` and `set query-param bar 2`.

**Comment by tlmii** [comment]
@ygalbel Now that #460 is in, which (by default) clears out headers and the path when running `connect`, we should update that in this PR to clear the global query parameters and add a `--persist-query-params` similar to the `--persist-headers` and `--persist-path` that were added in that PR.

File: src/Microsoft.HttpRepl/Commands/HelpCommand.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/Commands/HelpCommand.cs:189]: @ygalbel In the interest of moving this along (and since we won't push out a non-preview 5.1 release with this included for a while anyway) I'm going to go ahead and say we should go with what @atifaziz said: Make the command be `set query-param`. For now, lets leave it at that, without any additional `set query` or `set query-string` or aliases for any of it. We can see what the feedback is for the previews and get some more input from @bradygaster and others.

File: src/Microsoft.HttpRepl/HttpState.cs
Comment (tlmii) [reply src/Microsoft.HttpRepl/HttpState.cs:127]: @ygalbel I'll add a tracking issue for the `allowReserved` portion of this discussion and mention it here when I do. Then I think you just need the change @atifaziz mentioned suggested to encode the name as well as the value. A test would be good too :)

**Comment by ygalbel** [comment]
Don't remember where it stands.
Tag me if you need help from my side.

**Comment by tlmii** [comment]
We were actually just discussing this internally a couple of days ago. I'm going to try to pick it back up and get it into the product here in the next few days. It looks like I've let it go stale long enough I might need to do a fresh PR to get around some of the conflicts and changes that have happened in the interim. Regardless, should be moving forward soon.

**Comment by ygalbel** [comment]
Sure. Go ahead it's all yours.

**Comment by tlmii** [comment]
@ygalbel @bravecobra - OK, I've created a new PR: #611 that updates everything I could see that was unresolved above, renames a bunch of things, adds one more test, fixes an integration test that broke somewhere in the various .NET/kestrel upgrades but otherwise is fully the work above. Let me know over there what you think.

**Comment by tlmii** [comment]
Closing this since #611 went in. Thanks again @ygalbel for your help here!

**Pull request closed without merging**
